File: examples/swap.py
"""Quote ETH -> USDC (Base) through DarkRoute. Standard library only. Set RECIPIENT and REFUND_TO to create an order."""
import json, os, urllib.request

BASE = "https://app.darkroute.exchange/api/v1"
FROM = "nep141:eth.omft.near"
TO = "nep141:base-0x833589fcd6edb6e08f4c7c32d4f71b54bda02913.omft.near"
amount = os.environ.get("AMOUNT", "0.1")

def call(method, path, body=None):
    data = json.dumps(body).encode() if body else None
    req = urllib.request.Request(BASE + path, data=data, method=method, headers={"content-type": "application/json", "user-agent": "darkroute-example/1.0"})
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.load(r)

q = call("POST", "/quote", {"originAsset": FROM, "destinationAsset": TO, "amount": amount})
if not q.get("ok"): raise SystemExit(q.get("error"))
r = q["routes"][0]
print(f"quote: {amount} ETH -> {r['out']} USDC via {r['venue']} · fee {q['fee']['chargedPct']}% · all-in vs mid {q['cost']['allInPct']:.3f}%")

recipient, refund_to = os.environ.get("RECIPIENT"), os.environ.get("REFUND_TO")
if not recipient or not refund_to:
    print("set RECIPIENT and REFUND_TO to create an order"); raise SystemExit(0)
o = call("POST", "/order", {"originAsset": FROM, "destinationAsset": TO, "amount": amount, "recipient": recipient, "refundTo": refund_to})
if not o.get("ok"): raise SystemExit(o.get("error"))
s = call("GET", f"/order/{o['id']}")["order"]
print(f"order {o['id']}: {s['status']} · send {s['from']['amount']} {s['from']['symbol']} to {s['depositAddress']} before {s['deadline']}")

File: examples/swap.mjs
// Quote ETH -> USDC (Base) through DarkRoute. node >= 18. Pass RECIPIENT and REFUND_TO env to create an order.
const BASE = "https://app.darkroute.exchange/api/v1";
const FROM = "nep141:eth.omft.near";
const TO = "nep141:base-0x833589fcd6edb6e08f4c7c32d4f71b54bda02913.omft.near";
const amount = process.env.AMOUNT || "0.1";
const post = (path, body) => fetch(BASE + path, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) }).then((r) => r.json());

const q = await post("/quote", { originAsset: FROM, destinationAsset: TO, amount });
if (!q.ok) throw new Error(q.error);
const r = q.routes[0];
console.log(`quote: ${amount} ETH -> ${r.out} USDC via ${r.venue} · fee ${q.fee.chargedPct}% · all-in vs mid ${q.cost.allInPct?.toFixed(3)}%`);

const { RECIPIENT, REFUND_TO } = process.env;
if (!RECIPIENT || !REFUND_TO) { console.log("set RECIPIENT and REFUND_TO to create an order"); process.exit(0); }
const o = await post("/order", { originAsset: FROM, destinationAsset: TO, amount, recipient: RECIPIENT, refundTo: REFUND_TO });
if (!o.ok) throw new Error(o.error);
const s = await fetch(`${BASE}/order/${o.id}`).then((r) => r.json());
console.log(`order ${o.id}: ${s.order.status} · send ${s.order.from.amount} ${s.order.from.symbol} to ${s.order.depositAddress} before ${s.order.deadline}`);
